octree_web_viewer/client/main.ts

```diff
@@ -20,124 +20,203 @@ import { FirstPersonController } from './control';
+        );
         this.needsRender = true;
-      });
-    window.addEventListener('resize', () => this.onWindowResize(), false);
-    this.animate();
-  }
-
-  private onWindowResize() {
-    this.camera.aspect = window.innerWidth / window.innerHeight;
-    this.camera.updateProjectionMatrix();
-    this.renderer.setSize(window.innerWidth, window.innerHeight);
-    // Force a reload of the visible nodes.
-    this.lastFrustumUpdateTime = 0;
-  }
-
-  public animate() {
-    requestAnimationFrame(() => this.animate());
-
-    const time = performance.now();
-    if (this.controller.update()) {
-      this.lastMoveTime = time;
-      this.viewer.setMoving(true);
-      this.needsRender = true;
+        this.lastFrustumUpdateTime = 0;
+        this.lastMoveTime = 0;
+    }
+
+    private cleanup() {
+        // TODO (negin-z) block requests from the viewer that is going to be replaced
+        this.removeControls();
+        const renderArea = document.getElementById('renderArea');
+        if (this.renderer) {
+            renderArea.removeChild(this.renderer.domElement);
+            this.renderer.dispose();
+        }
+    }
+
+    private removeControls() {
+        if (this.guiRenderControls) {
+            this.gui.removeFolder(this.guiRenderControls);
+        }
+    }
+
+    private resetOctree() {
+        this.cleanup();
+        this.initCamera();
+        this.initScene();
+        this.initRenderer();
+        this.initOctreeViewer(this.octreeId);
+        this.addControls();
+    }
+
+    private setOctreeId = (newOctreeId: string) => {
+        this.octreeIdControl.setValue(newOctreeId);
+    }
+
+    private run = () => {
+        this.resetOctree();
+        this.animate();
     }
-    if (time - this.lastMoveTime > 250) {
-      this.viewer.setMoving(false);
-      this.needsRender = true;
+
+    public init() {
+        this.octreeId = "loading...";
+        this.gui = new GUI();
+
+        this.octreeIdControl =
+            this.gui
+                .add(this, 'octreeId')
+                .name('Point Cloud ID')
+                .onFinishChange(this.run);
+
+        //TODO (negin-z) error handling
+        this.fetchDefaultOctreeId()
+            .then(this.setOctreeId)
+            .then(this.run);
+
+        window.addEventListener('resize', () => this.onWindowResize(), false);
     }
-    if (
-      this.lastFrustumUpdateTime <= this.lastMoveTime &&
-      time - this.lastFrustumUpdateTime > 250
-    ) {
-      this.camera.updateMatrixWorld(false);
-      this.lastFrustumUpdateTime = time;
-      const matrix = new THREE.Matrix4().multiplyMatrices(
-        this.camera.projectionMatrix,
-        this.camera.matrixWorldInverse
-      );
-      this.viewer.frustumChanged(
-        matrix,
-        this.renderer.context.canvas.width,
-        this.renderer.context.canvas.height
-      );
+
+    private onWindowResize() {
+        this.camera.aspect = window.innerWidth / window.innerHeight;
```

Comment:
Why are you using here the window dimensions and not dimensions of `document.getElementById('renderArea')`?

Reply:
I do not know why this was used here.

Reply:
shall I put a todo here?